File: Euclid/Network/Codec/EncryptionDecoder.cs
using DotNetty.Buffers;
using DotNetty.Codecs;
using DotNetty.Transport.Channels;
using Euclid.Util;
using Euclid.Util.Encryption;
using System.Collections.Generic;

namespace Euclid.Network
{
    internal class EncryptionDecoder : ByteToMessageDecoder
    {
        private RC4 rc4;

        public EncryptionDecoder(RC4 rc4)
        {
            this.rc4 = rc4;
        }

        protected override void Decode(IChannelHandlerContext ctx, IByteBuffer buffer, List<object> output)
        {
            byte[] payload = new byte[buffer.ReadableBytes];
            buffer.ReadBytes(payload);

            var messagePayload = StringUtil.GetEncoding().GetString(payload);
            var decodedPayload = StringUtil.GetEncoding().GetBytes(this.rc4.Decipher(messagePayload));

            var result = Unpooled.Buffer();
            result.WriteBytes(decodedPayload);
            output.Add(result);
        }
    }
}

File: euclid/src/main/java/euclid/network/codec/EncryptionDecoder.java
package euclid.network.codec;

import euclid.util.StringUtil;
import euclid.util.encryption.RC4;
import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.ByteToMessageDecoder;

import java.util.List;

public class EncryptionDecoder extends ByteToMessageDecoder {

    private final RC4 rc4;

    public EncryptionDecoder(RC4 rc4) {
        this.rc4 = rc4;
    }

    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> output) {
        byte[] payload = new byte[buffer.readableBytes()];
        buffer.readBytes(payload);

        String messagePayload = new String(payload, StringUtil.getEncoding());
        byte[] decodedPayload = rc4.decipher(messagePayload).getBytes(StringUtil.getEncoding());

        ByteBuf result = Unpooled.buffer();
        result.writeBytes(decodedPayload);
        output.add(result);
    }
}
